src/modules/fw_lateral_longitudinal_control/FwLateralLongitudinalControl.cpp

```diff
@@ -0,0 +1,680 @@
+	_tecs.set_airspeed_filter_process_std_dev(_param_process_noise_standard_dev.get());
+
+	_performance_model.runSanityChecks();
+
+	_performance_model.updateParameters();
+	_roll_slew_rate.setSlewRate(radians(_param_fw_pn_r_slew_max.get()));
+
+	_tecs_alt_time_const_slew_rate.setSlewRate(TECS_ALT_TIME_CONST_SLEW_RATE);
+	_tecs_alt_time_const_slew_rate.setForcedValue(_param_fw_t_h_error_tc.get() * _param_fw_thrtc_sc.get());
+}
+
+void FwLateralLongitudinalControl::Run()
+{
+	if (should_exit()) {
+		_local_pos_sub.unregisterCallback();
+		exit_and_cleanup();
+		return;
+	}
+
+	perf_begin(_loop_perf);
+
+	/* only run controller if position changed */
+	// check for parameter updates
+	if (_parameter_update_sub.updated()) {
+		// clear update
+		parameter_update_s pupdate;
+		_parameter_update_sub.copy(&pupdate);
+
+		// update parameters from storage
+		parameters_update();
+	}
+
+	if (_local_pos_sub.update(&_local_pos)) {
+
+		const float control_interval = math::constrain((_local_pos.timestamp - _last_time_loop_ran) * 1e-6f,
+					       0.001f, 0.1f);
+		_last_time_loop_ran = _local_pos.timestamp;
+
+
+		_long_control_limits_sub.update();
+		_lateral_control_limits_sub.update();
+
+		_tecs.set_speed_weight(_long_control_limits_sub.get().speed_weight);
+		updateTECSAltitudeTimeConstant(checkLowHeightConditions()
+					       || _long_control_limits_sub.get().enforce_low_height_condition, control_interval);
+		_tecs.set_altitude_error_time_constant(_tecs_alt_time_const_slew_rate.getState());
+
+		if (_vehicle_air_data_sub.updated()) {
+			_air_density = PX4_ISFINITE(_vehicle_air_data_sub.get().rho) ? _vehicle_air_data_sub.get().rho : _air_density;
+			_tecs.set_max_climb_rate(_performance_model.getMaximumClimbRate(_air_density));
+			_tecs.set_min_sink_rate(_performance_model.getMinimumSinkRate(_air_density));
+		}
+
+		if (_vehicle_landed_sub.updated()) {
+			vehicle_land_detected_s landed{};
+			_vehicle_landed_sub.copy(&landed);
+			_landed = landed.landed;
+		}
+
+		_flight_phase_estimation_pub.get().flight_phase = flight_phase_estimation_s::FLIGHT_PHASE_UNKNOWN;
+
+		_vehicle_status_sub.update();
+		_control_mode_sub.update();
+		update_control_state();
+
+		if (_control_mode_sub.get().flag_control_manual_enabled && _control_mode_sub.get().flag_control_altitude_enabled
+		    && _local_pos.z_reset_counter != _z_reset_counter) {
+			if (_control_mode_sub.get().flag_control_altitude_enabled && _local_pos.z_reset_counter != _z_reset_counter) {
+				// make TECS accept step in altitude and demanded altitude
+				_tecs.handle_alt_step(_long_control_state.altitude_msl, _long_control_state.height_rate);
+			}
+		}
+
+		float pitch_sp{NAN};
+		float throttle_sp{NAN};
+
+		bool publish = false;
+
+		fw_longitudinal_control_setpoint_s longitudinal_sp = {empty_longitudinal_control_setpoint};
+		_fw_longitudinal_ctrl_sub.copy(&longitudinal_sp);
+
+		if (longitudinal_sp.timestamp > 0 && hrt_elapsed_time(&longitudinal_sp.timestamp) < 1_s) {
```

Comment:
I think that would need to be set better, Controller need to know if it is in charge or not. Like with the flags from the commander.